pkg/apis/environment/extension_view.go

```diff
@@ -54,3 +71,125 @@ type (
 
 	RouteGetResourceDefinitionsResponse = []*model.ResourceDefinitionOutput
 )
+
+type RouteStopRequest struct {
+	_ struct{} `route:"POST=/stop"`
+
+	model.EnvironmentQueryInput `path:",inline"`
+
+	stoppableResources model.Resources `json:"-"`
+}
+
+func (r *RouteStopRequest) Validate() error {
+	if err := r.EnvironmentQueryInput.Validate(); err != nil {
+		return err
+	}
+
+	resources, err := r.Client.Resources().Query().
+		Where(resource.EnvironmentID(r.ID)).
+		WithTemplate(func(tvq *model.TemplateVersionQuery) {
+			tvq.Select(
+				templateversion.FieldID,
+				templateversion.FieldTemplateID,
+				templateversion.FieldName,
+				templateversion.FieldVersion,
+				templateversion.FieldProjectID)
+		}).
+		WithProject(func(pq *model.ProjectQuery) {
+			pq.Select(project.FieldName)
+		}).
+		WithEnvironment(func(eq *model.EnvironmentQuery) {
+			eq.Select(environment.FieldName)
+		}).
+		All(r.Context)
+	if err != nil {
+		return err
+	}
+
+	stoppableResources := make(model.Resources, 0, len(resources))
+	stoppableResourceIDs := make([]object.ID, 0, len(resources))
+
+	for _, r := range resources {
+		if pkgresource.IsStoppable(r) {
```

Comment:
Are the Failed resources can be stopped?

Reply:
Update status validation for stop action